enhancements/separating-provider-specific-code-in-the-installer.md

```diff
@@ -0,0 +1,147 @@
+
+### Non-Goals
+
+Remove all provider-specific pieces from the code. That would require a complete refactor of the install config data structure and move the Terraform code. 
+
+## Proposal
+
+### Step 1: Create a Go interface
+
+We should create a Go interface that describes the functionality each provider must implement. For example, from the prototype:
+
+```go
+type PlatformV2 interface {
+	// Name returns the name of the platform.
+	Name() string
+
+	// Metadata translates the installConfig to the appropriate entry in clusterMetadata.
+	Metadata(clusterMetadata *types.ClusterMetadata, installConfig *installconfig.InstallConfig) error
+
+    //...
+}
+```
+
+Not all platforms implement all features. Optional features should be hidden behind a feature flag and a separate interface:
+
+```go
+var ErrNotAnIPIPlatform = errors.New("...")
+
+type PlatformV2 interface {
+    //...
+
+    // GetIPI returns the IPI-specific interface, or ErrNotAnIPIPlatform if the platform does not support IPI.
+	GetIPI() (IPIPlatformV2, error)
+}
+
+type IPIPlatformV2 interface {
+    // IPI-specific 
+}
+``` 
+
+### Step 2: Implement registry
+
+In this step we should implement a registry pattern where providers can register with their names.
+
+### Step 3: Implement feature-switch
+
+Once the registry is complete the installer core code must be changed. Provider-specific switch-case structures can be identified by searching for the usage of the `Name` constants in the `types` package. The existing switch-case should be wrapped in an if-else case. This will help with keeping the functionality of providers that have not yet switched.
+
+```go
+if v2, err := registry.Get(providerName); err == nil {
+    v2.CallSpecifiedFunction(...)
+} else if (!errors.Is(err, registry.ProviderNotFound)) {
+    return err
+} else {
+    switch (...) {
+        //...
+    }
+}
+```
+
+### Step 4: Migrate providers
+
+Providers need to be migrated to the new API.
+
+### Step 5: Remove fallback code
+
+Once all providers are migrated the fallback code can be completely removed.
+
+### Risks and Mitigations
+
+As with every refactor, this change contains the risk of breaking existing functionality. This can be mitigated by migrating only small portions of the code to the new API at a time.
+
+## Design Details
+
+### Open Questions
+
+1. How should the API look in detail?
+2. How should the implementation be merged? All-at-once, or bit by bit?
+
+### Test Plan
+
+Same as testing of the installer for regular releases.
+
+### Graduation Criteria
+
+TBD
+
+### Version Skew Strategy
+
+As the API develops it presents the problem of version skew. As new API calls are introduced they need to either be added to all providers. If that is not possible a new interface should be introduced and a compatibility wrapper should be added. Once all providers are migrated the temporary interface can be removed.
```

Comment:
This point is a big concern for me. Let's say that a provider needs a new parameter. Either we update every provider at once to add the new parameter. Or we support a new function with the new parameter. In the latter case, I would expect the providers that do not need the new parameter will **never** update to the new version of the function. Then, later, when a new parameter is added again, we get a third function. As time goes on, I could see every provider ending up with their own function. I look at the [machine assets](https://github.com/openshift/installer/blob/master/pkg/asset/machine) as an example. Every provider has a MachineSets and Machines function, but each one has its own unique signature.

Reply:
This issue has been addressed in the PoC (see below) I will update the document later.

Reply:
> As time goes on, I could see every provider ending up with their own function. I look at the [machine assets](https://github.com/openshift/installer/blob/master/pkg/asset/machine) as an example. Every provider has a MachineSets and Machines function, but each one has its own unique signature.

The parameters in an interface's function signature to generate an asset should just match the dependencies for that asset. Each platform/provider should have a signature that takes in all the dependencies and then returns whatever is needed to generate the asset (i.e. whatever is coming out of the switch statement). The PoC could be improved in this regard, I will leave a comment there.

Reply:
Actually, the machine config is pretty easy to standardize as far as I can tell. In my previous PoC I did this:

```go
	// GetMasterMachines returns a pool of machine configurations to create for the master nodes.
	GetMasterMachines(
		ic *types.InstallConfig,
		pool *types.MachinePool,
		rhcosImage *rhcos.Image,
		clusterID *installconfig.ClusterID,
	) ([]machineapi.Machine, error)

	// AddWorkerMachines adds a set of machines to the worker pool.
	AddWorkerMachines(
		ic *types.InstallConfig,
		pool *types.MachinePool,
		rhcosImage *rhcos.Image,
		clusterID *installconfig.ClusterID,
		machineSets []runtime.Object,
	) error
```

The API extraction is a rather simple process: there are a number of variables available there the provider-specific code is located and those make it into the API.